File: android/src/main/kotlin/com/example/multithread_downloads/DownloadUtils.kt
package com.example.multithread_downloads

import java.io.File
import java.net.URI

object DownloadUtils {

    fun extractFileName(url: String): String {
        return try {
            val uri = URI(url)
            val path = uri.path
            val fileName = path.substringAfterLast('/')
            if (fileName.isNotEmpty() && fileName.contains('.')) {
                fileName
            } else {
                "download_${System.currentTimeMillis()}.tmp"
            }
        } catch (e: Exception) {
            "download_${System.currentTimeMillis()}.tmp"
        }
    }

    fun isHlsUrl(url: String): Boolean {
        return url.endsWith(".m3u8", ignoreCase = true)
    }

    fun createDirectoryIfNotExists(path: String): Boolean {
        return try {
            val directory = File(path)
            if (!directory.exists()) {
                directory.mkdirs()
            } else {
                true
            }
        } catch (e: Exception) {
            false
        }
    }

    fun formatBytes(bytes: Long): String {
        val units = arrayOf("B", "KB", "MB", "GB", "TB")
        var size = bytes.toDouble()
        var unitIndex = 0

        while (size >= 1024 && unitIndex < units.size - 1) {
            size /= 1024
            unitIndex++
        }

        return String.format("%.2f %s", size, units[unitIndex])
    }

    fun formatSpeed(bytesPerSecond: Double): String {
        return "${formatBytes(bytesPerSecond.toLong())}/s"
    }

    fun calculateETA(remainingBytes: Long, speedBytesPerSecond: Double): String {
        if (speedBytesPerSecond <= 0) return "Unknown"

        val remainingSeconds = (remainingBytes / speedBytesPerSecond).toLong()
        val hours = remainingSeconds / 3600
        val minutes = (remainingSeconds % 3600) / 60
        val seconds = remainingSeconds % 60

        return when {
            hours > 0 -> String.format("%02d:%02d:%02d", hours, minutes, seconds)
            minutes > 0 -> String.format("%02d:%02d", minutes, seconds)
            else -> String.format("00:%02d", seconds)
        }
    }
}

File: ios/Classes/DownloadUtils.swift
import Foundation

struct DownloadUtils {

    static func extractFileName(url: String) -> String {
        do {
            guard let uri = URL(string: url) else {
                return "download_\(Int64(Date().timeIntervalSince1970 * 1000)).tmp"
            }

            let path = uri.path
            let fileName = String(path.split(separator: "/").last ?? "")

            if !fileName.isEmpty && fileName.contains(".") {
                return fileName
            } else {
                return "download_\(Int64(Date().timeIntervalSince1970 * 1000)).tmp"
            }
        } catch {
            return "download_\(Int64(Date().timeIntervalSince1970 * 1000)).tmp"
        }
    }

    static func isHlsUrl(url: String) -> Bool {
        return url.lowercased().hasSuffix(".m3u8")
    }

    static func createDirectoryIfNotExists(path: String) -> Bool {
        do {
            let directory = URL(fileURLWithPath: path)
            let fileManager = FileManager.default

            if !fileManager.fileExists(atPath: path) {
                try fileManager.createDirectory(at: directory, withIntermediateDirectories: true, attributes: nil)
            }
            return true
        } catch {
            return false
        }
    }

    static func formatBytes(bytes: Int64) -> String {
        let units = ["B", "KB", "MB", "GB", "TB"]
        var size = Double(bytes)
        var unitIndex = 0

        while size >= 1024 && unitIndex < units.count - 1 {
            size /= 1024
            unitIndex += 1
        }

        return String(format: "%.2f %@", size, units[unitIndex])
    }

    static func formatSpeed(bytesPerSecond: Double) -> String {
        return "\(formatBytes(bytes: Int64(bytesPerSecond)))/s"
    }

    static func calculateETA(remainingBytes: Int64, speedBytesPerSecond: Double) -> String {
        if speedBytesPerSecond <= 0 {
            return "Unknown"
        }

        let remainingSeconds = Int64(Double(remainingBytes) / speedBytesPerSecond)
        let hours = remainingSeconds / 3600
        let minutes = (remainingSeconds % 3600) / 60
        let seconds = remainingSeconds % 60

        switch true {
        case hours > 0:
            return String(format: "%02d:%02d:%02d", hours, minutes, seconds)
        case minutes > 0:
            return String(format: "%02d:%02d", minutes, seconds)
        default:
            return String(format: "00:%02d", seconds)
        }
    }
}
